lib/vmdb/plugins.rb

```diff
@@ -1,78 +1,77 @@
+require 'singleton'
+
 module Vmdb
   class Plugins
     include Singleton
+    include Enumerable
 
-    attr_reader :registered_ansible_content
-    attr_reader :registered_automate_domains
-
-    def initialize
-      @registered_automate_domains = []
-      @registered_provider_plugin_map = {}
-      @registered_ansible_content = []
-      @vmdb_plugins = []
-    end
-
-    def vmdb_plugins
-      @vmdb_plugins.empty? ? register_vmdb_plugins : @vmdb_plugins
+    def all
+      @all ||=
+        Rails::Engine.subclasses.select do |engine|
```

Comment:
Might be worth sticking a `require 'rails/engine'` at the top of this file.

Reasoning:  Technically for all of this to work, you still need to be within a `bundle` context of our application, which I think is basically assumed in all of the cases you have tested.  I think the error is easier to parse of `LoadError` versus `NameError` when this class might be loaded outside of a bundle context (by mistake or otherwise).

Reply:
interesting...I'll have to give that a try.

Reply:
This is what happened for me:

```ruby
$ ruby -r ./lib/vmdb/plugins.rb -e "puts Vmdb::Plugins.instance.all.inspect"
/Users/nicklamuro/code/redhat/manageiq/lib/vmdb/plugins.rb:10:in `all': uninitialized constant Vmdb::Plugins::Rails (NameError)
        from -e:1:in `<main>'
```

Though, I guess the opposite of what I was expecting happened when I added `require 'rails/engine'`:

```ruby
$ ruby -r ./lib/vmdb/plugins.rb -e "puts Vmdb::Plugins.instance.all.inspect"
[]
```

So maybe what you have is a better error.

Reply:
So, I did some more testing with the following commands:

```console
$ bundle exec ruby -e ...
$ irb
$ bundle exec irb
```

With and without the `require 'rails/engine'`.  Without, basically it barks because `Vmdb::Plugins::Rails` isn't defined, regardless if I am in a bundle context or not (a bit surprising).

But further more, because we autoload our paths and those aren't setup in these examples, the "works without rails" is a little misleading, because right now, with `require 'rails/engine'`, `Vmdb::Plugins.instance.all` returns `[]`.

Not sure how to fix that, and I would have to ponder what might actually be happening a bit further.  Also don't necessarily think this is show stoping, just something I was noticing when testing myself.

Reply:
I'd assume if you were trying to run this outside of rails, you'd have to manually require the various engines that you want to look at the various paths since we require the gems through `Bundler.require` [here](https://github.com/ManageIQ/manageiq/blob/ed49e069c2a6a3b5a2fdeca2cb9573035534ddcf/config/application.rb#L31).